rust/parsing.rs

```diff
@@ -0,0 +1,945 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} (Position: {})", self.message, self.index)
+    }
+}
+
+pub struct ParsedDateTime {
+    pub year: u32,
+    pub month: u32,
+    pub day: u32,
+    pub hour: u32,
+    pub minute: u32,
+    pub second: u32,
+    pub microsecond: u32,
+    pub offset: Option<i32>,
+    pub has_offset: bool,
+    pub tzname: Option<String>,
+    pub has_date: bool,
+    pub has_time: bool,
+    pub extended_date_format: bool,
+    pub time_is_midnight: bool,
+}
+
+impl ParsedDateTime {
+    pub fn new() -> ParsedDateTime {
+        ParsedDateTime {
+            year: 0,
+            month: 1,
+            day: 1,
+            hour: 0,
+            minute: 0,
+            second: 0,
+            microsecond: 0,
+            offset: None,
+            has_offset: false,
+            tzname: None,
+            has_date: false,
+            has_time: false,
+            extended_date_format: false,
+            time_is_midnight: false,
+        }
+    }
+}
+
+pub struct ParsedDuration {
+    pub years: u32,
+    pub months: u32,
+    pub weeks: u32,
+    pub days: u32,
+    pub hours: u32,
+    pub minutes: u32,
+    pub seconds: u32,
+    pub microseconds: u32,
+}
+
+impl ParsedDuration {
+    pub fn new() -> ParsedDuration {
+        ParsedDuration {
+            years: 0,
+            months: 0,
+            weeks: 0,
+            days: 0,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+            microseconds: 0,
+        }
+    }
+}
+
+pub struct Parsed {
+    pub datetime: Option<ParsedDateTime>,
+    pub duration: Option<ParsedDuration>,
+    pub second_datetime: Option<ParsedDateTime>,
+}
+
+impl Parsed {
+    pub fn new() -> Parsed {
+        Parsed {
+            datetime: None,
+            duration: None,
+            second_datetime: None,
+        }
+    }
+}
+
+pub struct Parser<'a> {
+    /// Input to parse.
+    src: &'a str,
+    /// Iterator used for getting characters from `src`.
+    chars: CharIndices<'a>,
+    /// Current byte offset into `src`.
+    idx: usize,
+    /// Current character
+    current: char,
+}
+
+impl<'a> Parser<'a> {
+    /// Creates a new parser from a &str.
+    pub fn new(input: &'a str) -> Parser<'a> {
+        let mut p = Parser {
+            src: input,
+            chars: input.char_indices(),
+            idx: 0,
+            current: '\0',
+        };
+        p.inc();
+        p
+    }
+
+    /// Increments the parser if the end of the input has not been reached.
+    /// Returns whether or not it was able to advance.
+    fn inc(&mut self) -> bool {
+        match self.chars.next() {
+            Some((i, ch)) => {
+                self.idx = i;
+                self.current = ch;
+                true
+            }
+            None => {
+                self.idx = self.src.len();
+                self.current = '\0';
+                false
+            }
+        }
+    }
```

Comment:
Returning `Option<char>` allows us to simplify code in some places (line 848 and 856), while not affecting current code as the `bool` isn't read anywhere.
```suggestion
    /// Returns the next character if it exists.
    fn inc(&mut self) -> Option<char> {
        if let Some((i, ch)) = self.chars.next() {
            self.idx = i;
            self.current = ch;
            Some(ch)
        } else {
            self.idx = self.src.len();
            self.current = '\0';
            None
        }
    }
```